server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoClusterHistogramAggregator.java

```diff
@@ -0,0 +1,296 @@
+                bucketOrd = -1 - bucketOrd;
+                clusters.get(bucketOrd).add(docInfo);
+            } else {
+                // Create a new cluster for this doc
+                clusters = context.bigArrays().grow(clusters, initialClusterOrds.size());
+                clusterCentroids = context.bigArrays().grow(clusterCentroids, initialClusterOrds.size());
+                ArrayList<DocInfo> cluster = new ArrayList<>();
+                cluster.add(docInfo);
+                clusters.set(
+                    initialClusterOrds.size()-1,
+                    cluster
+                );
+                clusterCentroids.set(initialClusterOrds.size()-1, docInfo.val);
+            }
+        } else {
+            // We already have k clusters. Find the nearest existing bucket to collect this doc into
+            int nearestBucket = getNearestCluster(docInfo.val);
+            assert nearestBucket != -1;
+
+            clusters.get(nearestBucket).add(docInfo);
+            int docCounts = bucketDocCount(nearestBucket);
+            double new_centroid = (clusterCentroids.get(nearestBucket) * docCounts + docInfo.val) / (docCounts + 1);
+            clusterCentroids.set(nearestBucket, new_centroid);
+        }
+    }
+
+    /**
+     * Returns the ordinal of the nearest bucket to val, or -1 if there are no buckets.
+     */
+    private int getNearestCluster(double val) {
+        double smallest_distance = Double.POSITIVE_INFINITY;
+        int nearest_bucket = -1;
+        for(int i = 0; i < k; i++){
+            double current_distance = Math.abs(clusterCentroids.get(i) - val);
+            if(current_distance < smallest_distance){
+                smallest_distance = current_distance;
+                nearest_bucket = i;
+            }
+        }
+        return nearest_bucket;
+    }
+
+    void recalculateCentroids(){
+        for(int i = 0; i < k; i++){
+            double sum = 0;
+            List<DocInfo> cluster = clusters.get(i);
+            for(DocInfo doc : cluster){
+                sum += doc.val;
+            }
+            clusterCentroids.set(i, sum / cluster.size());
+        }
+    }
+
+    void recalculateClusters(){
+        ObjectArray<List<DocInfo>> newClusters = context.bigArrays().newObjectArray(k);
+        for(int i = 0; i < k; i++){
+            newClusters.set(i, new ArrayList<>());
+        }
+
+        for(int oldCluster = 0; oldCluster < k; oldCluster++){
+            List<DocInfo> cluster = clusters.get(oldCluster);
+            for(DocInfo docInfo : cluster){
+                int newCluster = getNearestCluster(docInfo.val);
+                newClusters.get(newCluster).add(docInfo);
+            }
+        }
+        Releasables.close(clusters);
+        clusters = newClusters;
+    }
+
+    public void kMeansIteration(){
+        recalculateCentroids();
+        recalculateClusters();
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long bucket) throws IOException {
```

Comment:
Since this is deferring collection of sub aggregations until after clustering, I think we will need to collect sub aggs in this method somewhere (similar to how the `DeferringBucketCollector`s work, or by implementing this as a deferred agg).

Reply:
Hmm. I thought that sub-aggregations aren't run on a document until we collect it into a bucket. And since we only actually call `collectBucket()` after we run k-means, the sub aggregations would automatically be deferred until this point.

I can definitely update this class to be a deferred buckets aggregator. But I'm curious when the sub aggregations would be called in the current code. During the collection phase?